### ISSUE TITLE
netatalk2: Installing ELF binaries into /etc?

### ISSUE THREAD

**Issue by VorpalBlade** [Netatalk/netatalk#407]
Arch Linux has a package linting tool ([namcap](https://man.archlinux.org/man/namcap.1)) which I ran on the package I'm creating for Arch Linux AUR. Normally these are packaging issues, but I believe I found some upstream issues as well. This is one of those:

```
netatalk2 E: ELF file ('etc/netatalk/uams/uams_dhx2_pam.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_dhx2_passwd.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_dhx_pam.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_dhx_passwd.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_guest.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_pam.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_passwd.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_pgp.so') outside of a valid path.
netatalk2 E: ELF file ('etc/netatalk/uams/uams_randnum.so') outside of a valid path.
```

I'm not sure what the intention is here. Wouldn't it make more sense to install these somewhere under PREFIX and install symlinks under /etc? Note that the `E:` in the output signifies an error in namcap.

**Comment by VorpalBlade** [comment]
Aha, it seems like `--enable-fhs` fixes this (and changes some other things too). That was not obvious. And I don't think installing these under `/etc` is a sane default to be honest.

**Comment by rdmark** [comment]
I can't see how pkgconfdir would be a good default location for the UAM libraries, either. Did some code archaeology and found this was introduced in 2005 when netatalk 2.0 was being developed (AFP3.x support). https://github.com/Netatalk/netatalk/commit/ecfc96169ab669b578e53fa8e13592934fe37788#diff-bb21aa33a3f69ccb36c68b220f40ad08f29b9cd2c05dfedae7b9e3d5d4d08f6b

This autoconf check is to blame I think. https://github.com/Netatalk/netatalk/blob/310ac91641c33913f6d9fd1db5fa63145749bb5c/configure.ac#L502

```
AC_ARG_WITH(uams-path,
	[  --with-uams-path=PATH   path to UAMs [[PKGCONF/uams]]],[
		uams_path="$withval"
	],[
		uams_path="${PKGCONFDIR}/uams"
	]
)
```

I would be supportive of making the default `uams_path="${libdir}/netatalk"` instead.

**Comment by rdmark** [comment]
This is probably the relevant default in netatalk3: https://github.com/Netatalk/netatalk/blob/557d0bf2f6e9274c3a454a9f9ab1774e171910a6/configure.ac#L146

**Comment by rdmark** [comment]
For the scope of this ticket, I plan to make it the goal to harmonize the behavior of --enable-fhs between netatalk2 and 3.

**Comment by rdmark** [comment]
In netatalk3, this is when the change was made from pkgconfdir to libdir (March 2012): https://github.com/Netatalk/netatalk/commit/ed9a339b481cbe160d08aa2cd0a6d76307d29dcf

**Comment by rdmark** [comment]
This should do the trick I think https://github.com/Netatalk/netatalk/pull/420

You should no longer have to use `--enable-fhs` to get a reasonable default for the UAM libs.

**Comment by rdmark** [comment]
BTW, some of the default configuration changes between netatalk2 and 3 are documented in https://netatalk.sourceforge.io/3.1/htmldocs/upgrade.html

**Comment by VorpalBlade** [comment]
I'm not sure what the intent of `--enable-fhs` is, but the fact that it now prefers `/bin`, `/lib` etc over `/usr/bin`, `/usr/lib` and so on is problematic for all those distros that have done the usr-merge. Which is everything except Debian as I understand it.

In addition, Arch doesn't use `/sbin` or `/usr/sbin`, instead preferring to put everything into `/usr/bin`.

If I understand autoconf correctly (which is NOT a given!) this will override any `--prefix`, `--sbindir` etc the user sets as well? (Or is that processed afterwards?) So it is now not a usable option for me on Arch.

**Comment by rdmark** [comment]
Good points. I'm diving into the ML archives now to see if I find some clues to why --enable-fhs was designed the way it was.

For starters, here's someone who calls out the same issue about --enable-fhs overriding all other user defined options... in 2001 ;) https://sourceforge.net/p/netatalk/mailman/netatalk-devel/thread/20011123161326.A10389%40jroger.in-berlin.de/#msg9117928

This is another pertinent thread from 2004 https://sourceforge.net/p/netatalk/mailman/netatalk-devel/thread/BDB2F47C.3D521%25Thomas.Kaiser%40phg-online.de/#msg12928195

To summarize, the stated intention of --enable-fhs is to have a shorthand option for packagers, while keeping the default behavior appropriate for ad-hoc installation (i.e. $prefix = "/"  vs. $prefix = "/usr/local" etc)

With this in mind let's revisit the changes we want to make! Keep in mind that in addition to Linux, we support OpenSolaris and the BSDs as well.

**Comment by rdmark** [comment]
So let's revert back to solving the narrow problem of default UAM library path in this ticket. I force pushed a change to my PR to make it so.

@VorpalBlade May I ask you to create a new ticket and share what you think are good defaults for package management purposes in this day and age? As we have concluded before, the "FHS compatibility" settings were decided 10-20 years ago and should be revisited.

**Comment by VorpalBlade** [comment]
> May I ask you to create a new ticket and share what you think are good defaults for package management purposes in this day and age?

Very good question, I'm not an expert as I only package some stuff for Arch Linux AUR (Arch User Repository). I'm by no means a "professional packager" (if there is such a thing). I'll think about it.

**Comment by rdmark** [comment]
This should fix the tests https://github.com/Netatalk/netatalk/pull/522

**Comment by VorpalBlade** [comment]
@rdmark Since I didn't get a response when asking this in the PR:

Will this fix still place the UAMS in a sensible location on Linux? Or will they start going into /etc again?
If so: What flags do I need to append to configure to make it sensible again?

**Comment by rdmark** [comment]
Err, I did respond over there. As I said above this is for the test suite, nothing else.

**Comment by VorpalBlade** [comment]
> Err, I did respond over there. As I said above this is for the test suite, nothing else.

Sorry, see it now. Not sure how I missed it before.

**Comment by rdmark** [comment]
The test suite is now using dummy paths. It doesn't really use them.
